### PR TITLE
ADD: add getEdit implements and CommitInfoParser

### PR DESCRIPTION
**Pull request by sunchenglong**

### CODE REVIEW

File: src/main/java/com/google/gerrit/extensions/common/EditInfo.java [hunk 1]
```diff
@@ -14,10 +14,13 @@
 
 package com.google.gerrit.extensions.common;
 
+import com.google.gson.annotations.SerializedName;
+
 import java.util.Map;
 
 public class EditInfo {
   public CommitInfo commit;
+  @SerializedName("baseRevision")
```
Comment (uwolfer) [inline src/main/java/com/google/gerrit/extensions/common/EditInfo.java]: Why is this required?

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/ChangeApiRestClient.java [hunk 2]
```diff
@@ -205,6 +209,12 @@ public ChangeInfo info() throws RestApiException {
         return get(EnumSet.noneOf(ListChangesOption.class));
     }
 
+    public EditInfo getEdit() throws RestApiException {
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/ChangeApiRestClient.java]: Can you please add a test for this as well?

Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/ChangeApiRestClient.java]: Missing @Override annotation.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/ChangesRestClient.java [hunk 3]
```diff
@@ -40,21 +40,24 @@
...
 
     public ChangesRestClient(GerritRestClient gerritRestClient,
                              ChangesParser changesParser,
                              CommentsParser commentsParser,
                              FileInfoParser fileInfoParser,
                              DiffInfoParser diffInfoParser,
                              SuggestedReviewerInfoParser suggestedReviewerInfoParser,
-                             ReviewerInfoParser reviewerInfoParser){
+                             ReviewerInfoParser reviewerInfoParser,
+                             EditInfoParser editInfoParser){
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/ChangesRestClient.java]: Please add a space between ) and { (I see, it was not there before already ;) ).

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+public class CommitInfoParser {
+    private static final Type TYPE = new TypeToken<List<CommitInfoParser>>() {}.getType();
+
+    private final Gson gson;
+
+    public CommitInfoParser(Gson gson) {
+        this.gson = gson;
+    }
+
+    public List<CommitInfo> parseCommitInfos(JsonElement result) throws RestApiException {
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java]: `throws RestApiException` not necessary.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+public class EditInfoParser {
+    private static final Type TYPE = new TypeToken<List<EditInfoParser>>() {}.getType();
+
+    private final Gson gson;
+
+    public EditInfoParser(Gson gson) {
+        this.gson = gson;
+    }
+
+    public List<EditInfo> parseEditInfos(JsonElement result) throws RestApiException {
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java]: `throws RestApiException` not necessary.

File: src/test/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParserTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+public class EditInfoParserTest extends AbstractParserTest {
+    private final EditInfoParser editInfoParser = new EditInfoParser(getGson());
+
+    @Test
+    public void testParseEditInfo() throws Exception {
+        JsonElement jsonElement = getJsonElement("edit.json");
+        List<EditInfo> editInfos = editInfoParser.parseEditInfos(jsonElement);
+        Truth.assertThat(editInfos).hasSize(1);
+        System.out.println(editInfos.get(0));
```
Comment (uwolfer) [inline src/test/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParserTest.java]: Please remove this line.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        if (!result.isJsonArray()) {
+            return Collections.singletonList(gson.fromJson(result,  CommitInfo.class));
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java]: There are two spaces after the comma.

File: src/main/java/com/google/gerrit/extensions/common/EditInfo.java (hunk 1 above)
Comment (sunchenglong) [inline src/main/java/com/google/gerrit/extensions/common/EditInfo.java]: Because the key is "baseRevision" , the "R" is upcase, but gson cannot process it.

Comment (uwolfer) [inline src/main/java/com/google/gerrit/extensions/common/EditInfo.java]: The naming of `baseRevision` got changed to `base_revision` in Gerrit itself:
https://gerrit-review.googlesource.com/c/67956/

You can also see the updated JSON response for your test here:
https://gerrit-review.googlesource.com/Documentation/rest-api-changes.html#get-edit-detail

Does this also work with your version of Gerrit? In that case, SerializedName can be removed. Otherwise, it would be good to have two tests, one for the old response and one for the new one.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.urswolfer.gerrit.client.rest.http.changes;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gerrit.extensions.common.CommitInfo;
+import com.google.gerrit.extensions.restapi.RestApiException;
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java]: This can be removed as well.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.urswolfer.gerrit.client.rest.http.changes;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gerrit.extensions.common.EditInfo;
+import com.google.gerrit.extensions.restapi.RestApiException;
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java]: This can be removed as well.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Chenglong Sun
+ */
+public class CommitInfoParser {
+    private static final Type TYPE = new TypeToken<List<CommitInfoParser>>() {}.getType();
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/CommitInfoParser.java]: I think this should be `List<CommitInfo>`.

File: src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * @author Chenglong Sun
+ */
+public class EditInfoParser {
+    private static final Type TYPE = new TypeToken<List<EditInfoParser>>() {}.getType();
```
Comment (uwolfer) [inline src/main/java/com/urswolfer/gerrit/client/rest/http/changes/EditInfoParser.java]: And here `List<EditInfo>`.

File: src/main/java/com/google/gerrit/extensions/common/EditInfo.java (hunk 1 above)
Comment (sunchenglong) [inline src/main/java/com/google/gerrit/extensions/common/EditInfo.java]: I have used version v2.12.2,  https://gerrit.wikimedia.org/r/Documentation/rest-api-changes.html.

File: src/main/java/com/google/gerrit/extensions/common/EditInfo.java [hunk 12]
```diff
@@ -18,7 +18,7 @@
 
 public class EditInfo {
   public CommitInfo commit;
-  public String baseRevision;
+  public String base_revision;
```
Comment (uwolfer) [inline src/main/java/com/google/gerrit/extensions/common/EditInfo.java]: I don't think this change is required? It should convert fine now also with camel case notation?

**Comment by uwolfer** [comment]
@sunchenglong: Thanks for your work! :)

**Comment by sunchenglong** [comment]
@uwolfer You are welcome!
